File: leetcode-js/275-h-index-ii.js
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function(citations) {
    for (let i = citations.length - 1; i >= 0; i -= 1) {
        if (citations[i] >= citations.length - i) {
            if (i > 0) {
                if (citations[i-1] <= citations.length - i) {
                    return citations.length - i;
                }
            } else {
                return citations.length;
            }
        } else {
            break;
        }
    }

    return 0;
};

File: leetcode-java/275-h-index-ii.java
public class Solution {
    public int hIndex(int[] citations) {
        int left = 0;
        int right = citations.length - 1;
        int len = citations.length;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (citations[mid] == len - mid) {
                return citations[mid];
            } else if (citations[mid] > len - mid) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return len - (right + 1);
    }
}
